src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -14635,29 +14706,637 @@ public S(A a, B b)
         {
-            CSharpCompilation c = CreateStandardCompilation(new [] { attributesDefinitions, @"
-using nullableString = System.String?;
+            var source =
+@"class Person
+{
+    internal string FirstName { get; set; }
+    internal string LastName { get; set; }
+    internal string? MiddleName { get; set; }
+}
+class Program
+{
+    static void F(Person p)
+    {
+        G(p.MiddleName);
+        if (p.MiddleName != null) G(p.MiddleName);
+    }
+    static void G(string s)
+    {
+    }
+}";
+            var comp = CreateStandardCompilation(
+                new[] { source, attributesDefinitions },
+                parseOptions: TestOptions.Regular.WithNullCheckingFeature(flags: NullableReferenceFlags.None));
+            comp.VerifyDiagnostics(
+                // (11,11): warning CS8604: Possible null reference argument for parameter 's' in 'void Program.G(string s)'.
+                //         G(p.MiddleName);
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "p.MiddleName").WithArguments("s", "void Program.G(string s)").WithLocation(11, 11));
+        }
 
-class C
+        [Fact]
+        public void LocalTypeInference()
+        {
+            var source =
+@"class C
 {
-    static void Main()
+    static void F(string? s, string? t)
+    {
+        if (s != null)
+        {
+            var x = s;
+            G(x); // no warning
+            x = t;
+        }
+        else
+        {
+            var y = s;
+            G(y); // warning
+            y = t;
+        }
+    }
+    static void G(string s)
+    {
+    }
+}";
+            var comp = CreateStandardCompilation(
+                new[] { source, attributesDefinitions },
+                parseOptions: TestOptions.Regular.WithNullCheckingFeature(flags: NullableReferenceFlags.None));
+            comp.VerifyDiagnostics(
+                // (14,15): warning CS8604: Possible null reference argument for parameter 's' in 'void C.G(string s)'.
+                //             G(y); // warning
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "y").WithArguments("s", "void C.G(string s)").WithLocation(14, 15));
+        }
+
+        [Fact]
+        public void AssignmentInCondition_01()
+        {
+            var source =
+@"class C
+{
+    object P => null;
+    static void F(object o)
+    {
+        C? c;
+        while ((c = o as C) != null)
+        {
+            o = c.P;
+        }
+    }
+}";
+            var comp = CreateStandardCompilation(
+                new[] { source, attributesDefinitions },
+                parseOptions: TestOptions.Regular.WithNullCheckingFeature(flags: NullableReferenceFlags.None));
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void AssignmentInCondition_02()
+        {
+            var source =
+@"class C
+{
+    object? P => null;
+    static void F(object? o)
+    {
+        C? c;
+        while ((c = o as C) != null)
+        {
+            o = c.P;
```

Comment:
`DataFlowPass.VisitBinaryOperatorChildren` executes `StateWhenTrue[MakeSlot(operandComparedToNull)] = true;` and `DataFlowPass.MakeSlot()` handles `AssignmentOperator` at line 933.